### PR TITLE
Dynamic function replacement

### PR DESCRIPTION
**Pull request by aschwaighofer**
Implement dynamic function replacement as described in https://forums.swift.org/t/dynamic-method-replacement.

Allow ``dynamic`` on non-``@objc`` classes, ``struct``, and `enum` functions, properties, initializers.

Dynamic replacements are currently expressed by adding the ``@_dynamicReplacement(for:)`` attribute on functions defined in an extensions of the type or at the top-level.

```
// Module A
struct Foo {
 dynamic func bar() {}
}
```

```
// Module B
extension Foo {
  @_dynamicReplacement(for: bar()0
  func barReplacement() {
    ...
    // Calls previously active implementation of bar()
    bar()
  }
```

Dynamic replacements become active at load time of the module containing the replacement.

The runtime implementation allows for an implementation in the future where dynamic replacements can be grouped in an 'dynamic replacement' scope and the scope can be dynamically enabled/disabled.

Dynamic replacements chain such that a dynamic replacement can call the previously active implementation of the dynamically replaced function.

```
dynamic_replacement_scope AGroupOfReplacements {
   extension Foo {
     func replacedFunc() {}
   }
   extension AnotherType {
     func replacedFunc() {}
   }
}

AGroupOfReplacements.enable()
...
AGroupOfReplacements.disable()
```

### CODE REVIEW

File: lib/AST/ASTDumper.cpp [hunk 1]
```diff
@@ -768,8 +768,9 @@ namespace {
         OS << " final";
       if (VD->isObjC())
         OS << " @objc";
-      if (VD->isDynamic())
+      if (VD->isDynamic()) {
```
Comment (slavapestov) [inline lib/AST/ASTDumper.cpp]: Is this edit necessary?

File: lib/SIL/SILVerifier.cpp [hunk 2]
```diff
@@ -1402,6 +1402,14 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
 
     SILFunction *RefF = FRI->getReferencedFunction();
 
+    if (isa<FunctionRefInst>(FRI))
+      require(!RefF->isDynamicallyReplaceable(), "function_ref cannot reference a [dynamically_replaceable] function");
```
Comment (slavapestov) [inline lib/SIL/SILVerifier.cpp]: These lines are too long

**Comment by slavapestov** [comment]
How does it work if a class method is dynamically replaceable? Is it still called via class_method and then we have to emit a vtable thunk which does the dynamic_function_ref?

**Comment by aschwaighofer** [comment]
```
$ cat C.swift
class Foo {
  dynamic func bar() {}
}
```

```
// Foo.bar()
sil hidden [dynamically_replacable] @$s1C3FooC3baryyF : $@convention(method) (@guaranteed Foo) -> () {
// %0                                             // user: %1
bb0(%0 : @guaranteed $Foo):
  debug_value %0 : $Foo, let, name "self", argno 1 // id: %1
  %2 = tuple ()                                   // user: %3
  return %2 : $()                                 // id: %3
}

sil_vtable Foo {
  #Foo.bar!1: (Foo) -> () -> () : @$s1C3FooC3baryyF     // Foo.bar()
...
```

**Comment by aschwaighofer** [comment]
The [dynamically_replaceable] SIL function does the dispatch:

```
@"$s1C3FooC3baryyFTX" = hidden global %swift.dyn_repl_link_entry { i8* bitcast (void (%T1C3FooC*)* @"$s1C3FooC3baryyFTI" to i8*), %swift.dyn_repl_link_entry* null }, align 8

define hidden swiftcc void @"$s1C3FooC3baryyF"(%T1C3FooC* swiftself) #0 {
entry:
  %1 = load i8*, i8** getelementptr inbounds (%swift.dyn_repl_link_entry, %swift.dyn_repl_link_entry* @"$s1C3FooC3baryyFTX", i32 0, i32 0), align 8
  %2 = bitcast i8* %1 to void (%T1C3FooC*)*
  call swiftcc void %2(%T1C3FooC* swiftself %0)
  ret void
}


define hidden swiftcc void @"$s1C3FooC3baryyFTI"(%T1C3FooC* swiftself) #1 {
entry:
  %self.debug = alloca %T1C3FooC*, align 8
  %1 = bitcast %T1C3FooC** %self.debug to i8*
  call void @llvm.memset.p0i8.i64(i8* align 8 %1, i8 0, i64 8, i1 false)
  store %T1C3FooC* %0, %T1C3FooC** %self.debug, align 8
  ret void
}
```

**Comment by aschwaighofer** [comment]
dynamic_function_ref is different to function_ref only in that it is verified that we call a SIL function that is [dynamically_replaceable].

**Comment by aschwaighofer** [comment]
To answer your question: yes we still dispatch via class_method to the callee.

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please clean test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test os x platform

**Comment by aschwaighofer** [comment]
@swift-ci Please test source compatibility

**Comment by aschwaighofer** [comment]
The source compatibility failure is the same that also occurs on trunk.

**Comment by Whirlwind** [comment]
Hi, I could not work on the scope:
```
dynamic_replacement_scope AGroupOfReplacements {
   extension Foo {
     func replacedFunc() {}
   }
   extension AnotherType {
     func replacedFunc() {}
   }
}

AGroupOfReplacements.enable()
...
AGroupOfReplacements.disable()
```
Is it public?

By the way, if there are two replacement methods to replace the original method, how to call the chain?
```
public class Core {
    public dynamic func originalFunc() { print("core") }
}

extension Core {
    @_dynamicReplacement(for: originalFunc())
    public func a() { 
         print("a")
         originalFunc()
    }
}

extension Core {
    @_dynamicReplacement(for: originalFunc())
    public func b() { 
         print("b")
         originalFunc()
    }
}
```
I hope to get the 
```
a
b
core
```
Or, get the 
```
b
a
core
```
But now, I get the 
```
b
core
```
Lost the `a`.

**Comment by Changzw** [comment]
can I use it to class method?
